### PR TITLE
Flow Chat Engine

### PR DESCRIPTION
**Pull request by Falven**
# Description

Introduces Flow Chat engine, a chat engine better optimized for interactions within smaller context sizes.
Imagine a conversation history using the "context" chat engine with GPT-3.5 4K as follows:

```
> |---------|
> | Sys     |
> | Context |
> |---------|
> | User    |
> |---------|
> | LLM     |
> |---------|
```

When necessitating a prompt that covers crucial LLM instructions on, say the way a document should be interpreted, the "context" and "context condensed" Chat Engines quickly falls short. Observe as we add more to the conversation, below, the important instructions and context fall outside of the context window.

```
|---------|
| Sys     |
| Context |
|---------|
> | User    |
> |---------|
> | LLM     |
> |---------|
> | User    |
> |---------|
> | LLM     |
> |---------|
```

The Flow chat engine ensures that the instructions are always included within the context window by including them as part of a user message prompt. Additionally, it condenses the chat history to maintain it in simple QA format.

```
|---------|
| Sys     |
|---------|
| User    |
|---------|
> | LLM     |
> |---------|
> | User    |
> | Context |
> |---------|
> | LLM     |
> |---------|
```

This Chat Engine is particularly useful for scenarios where a simple QA is needed using LLMs that do not support function calling and have small context sizes. As is a typical industry case for cheap Document QAs using the cheapest GPT 3.5 4K fine tunable model.

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [X] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [X] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] Added new unit/integration tests
- [X] Added new notebook (that tests end-to-end)
- [X] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [] I have added Google Colab support for the newly added notebooks.
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

**Comment by logan-markewich** [comment]
FYI just to clarify, your illustration is not quite correct. The system/context is included no matter what, but the buffer on the chat history will move (the context is in the prefix variables, its always added)
https://github.com/run-llama/llama_index/blob/577a81d4c5beb42f27988c5def2d8913678510a9/llama-index-core/llama_index/core/chat_engine/context.py#L169

File: llama-index-core/llama_index/core/chat_engine/flow.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+        )
+
+    @trace_method("chat")
+    def chat(
+        self, message: str, chat_history: Optional[List[ChatMessage]] = None
+    ) -> AgentChatResponse:
+        if chat_history is not None:
+            self._memory.set(chat_history)
+
+        qa_context_str_template, nodes = self._generate_qa_context(message)
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/chat_engine/flow.py:146]: Do you see any issues with only using the most recent message to retrieve? (tbh, this was kind of the key reason for introducing the condense_plus_context chat engine. Maybe we can bundle similar logic here under a boolean and avoid yet another class for this mode?)

File: llama-index-core/llama_index/core/chat_engine/flow.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+            )
+        )
+        all_messages = self._prefix_messages + self._memory.get(
+            initial_token_count=prefix_messages_token_count
+        )
+        chat_response = self._llm.chat(all_messages)
+        ai_message = chat_response.message
+        self._memory.put(ai_message)
+
+        self._memory.get_all()[-2].content = message
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/chat_engine/flow.py:165]: This feels very... weird? What is happening here? Can this be done before putting things in memory? If not, this needs a comment explaining

**Review by logan-markewich — COMMENTED** [review]
This looks mostly good, just a few comments

**Comment by Falven** [comment]
> FYI just to clarify, your illustration is not quite correct. The system/context is included no matter what, but the buffer on the chat history will move (the context is in the prefix variables, its always added)
> 
> https://github.com/run-llama/llama_index/blob/577a81d4c5beb42f27988c5def2d8913678510a9/llama-index-core/llama_index/core/chat_engine/context.py#L169

I see, so the prefix_messages are always included (which includes the system message with the context) and then the conversation history (SYSTEM and USER messages) are truncated depending on the configured token count.

CleanShot 2024-04-02 at 00 30 14@2x

I think that observation renders this PR useless. ;)

**Comment by logan-markewich** [comment]
@Falven well, not useless per say. This pr is  just an inverted context chat engine right now 👀 it could honestly be a mode on the two current chat engines?

**Comment by Falven** [comment]
> @Falven well, not useless per say. This pr is  just an inverted context chat engine right now 👀 it could honestly be a mode on the two current chat engines? 

Well it's my understanding that System messages have higher precedence than user messages so I can't imagine why anyone would want this if the system message (and therefore instructions) is always included.

**Comment by logan-markewich** [comment]
🤷🏻 not ever model supports system messages. But sure. Wi close this out

**Pull request closed without merging**